source/extensions/filters/http/cache/http_cache.h

```diff
@@ -241,6 +241,14 @@ class InsertContext {
   // Inserts trailers into the cache.
   virtual void insertTrailers(const Http::ResponseTrailerMap& trailers) PURE;
 
+  // This routine is called prior to an InsertContext being destroyed. InsertContext is responsible
+  // for making sure that any async events are cleaned up in the context of this routine. This
+  // includes timers, network calls, etc. The reason there is an onDestroy() method vs. doing this
+  // type of cleanup in the destructor is due to the deferred deletion model that Envoy uses to
+  // avoid stack unwind complications. InsertContext must not invoke any callbacks after having
+  // onDestroy() invoked.
+  virtual void onDestroy() PURE;
```

Comment:
I'm not completely sure why we need onDestroy() here and below. This should never be called in the context of a cache callback, so why can't it be handled by the relevant destructors?

Reply:
My design originally didn't include (Insert|Lookup)Context::onDestroy, but I discovered that it was necessary in order to avoid potential data races. Without onDestroy, consider the following sequence with a hypothetical plugin-provided RpcInsertContext that implements InsertContext:

1. [Filter's thread] CacheFilter calls InsertContext::insertBody.
2. [Filter's thread] RpcInsertContext sends RPC and returns.
3. [Filter's thread] Client disconnects; CacheFilter dtor begins
4. [Filter's thread] RpcInsertContext dtor begins
5. [Other thread] RPC completes and calls RpcInsertContext::onRpcDone.
RpcInsertContext's dtor and onRpcDone cause a data race in RpcInsertContext.

There are several ways we could prevent this:
1. Document that all access to classes that inherit from InsertContext must be on the Filter's thread. This complicates the requirements on plugins. Mistakes will be painful. I recommend against this approach.
2. Share ownership of InsertContext. Some plugins would give shared_ptrs to their callbacks (to keep InsertContext alive). Plugins that want to cancel aync operations would only hold weak_ptrs. When CacheFilter gets deleted and drops its InsertContextSharedPtr, RpcInsertContext's dtor would cancel any outstanding async operations. Anything that completes after that would bail when it fails to convert the weak_ptr to a shared_ptr.
3. Call InsertContext::onDestroy in CacheFilter::onDestroy. RpcInsertContext::onDestroy cancels any outstanding async operations. If necessary, it blocks on that cancellation.

I chose onDestroy because an earlier shared ownership approach received unfavorable reviews, and because onDestroy is an established Envoy idiom. I consider both shared ownership and onDestroy to be reasonable approaches. onDestroy leaves more room for implementation mistakes, but it's more explicit, and avoids shared ownership.

Reply:
> I chose onDestroy because an earlier shared ownership approach received unfavorable reviews, and because onDestroy is an established Envoy idiom. I consider both shared ownership and onDestroy to be reasonable approaches. onDestroy leaves more room for implementation mistakes, but it's more explicit, and avoids shared ownership.

Right, OK. I vaguely recall us having a conversation on this last time. Sorry for making you type this out again. Would it be worth it to take this nice detailed comment and put it in the code somewhere so we have this history and I don't end up asking this again in a few months? I think it's worthwhile, even if slightly slimmed down. Up to you.

Reply:
@toddmgreer Happy to make any additions/modifications to these functions documentation as you see fit.